pkg/ugrpc/c1_credential_provider.go

```diff
@@ -0,0 +1,245 @@
+	Expiry      int    `json:"expires_in"`
+}
+
+type c1TokenSource struct {
+	clientID     string
+	clientSecret *jose.JSONWebKey
+	tokenHost    string
+	httpClient   *http.Client
+}
+
+func parseClientID(input string) (string, string, error) {
+	// split the input into 2 parts by @
+	items := strings.SplitN(input, "@", 2)
+	if len(items) != 2 {
+		return "", "", ErrInvalidClientID
+	}
+	clientName := items[0]
+
+	// split the right part into 2 parts by /
+	items = strings.SplitN(items[1], "/", 2)
+	if len(items) != 2 {
+		return "", "", ErrInvalidClientID
+	}
+
+	return clientName, items[0], nil
+}
+
+func parseSecret(input []byte) (*jose.JSONWebKey, error) {
+	items := bytes.SplitN(input, []byte(":"), 4)
+	if len(items) != 4 {
+		return nil, ErrInvalidClientSecret
+	}
+
+	jwkData, err := base64.RawURLEncoding.DecodeString(string(items[3]))
+	if err != nil {
+		return nil, ErrInvalidClientSecret
+	}
+
+	npk := &jose.JSONWebKey{}
+	err = npk.UnmarshalJSON(jwkData)
+	if err != nil {
+		return nil, ErrInvalidClientSecret
+	}
+
+	if npk.IsPublic() || !npk.Valid() {
+		return nil, ErrInvalidClientSecret
+	}
+
+	_, ok := npk.Key.(ed25519.PrivateKey)
+	if !ok {
+		return nil, ErrInvalidClientSecret
+	}
+
+	return npk, nil
+}
+
+func (c *c1TokenSource) Token() (*oauth2.Token, error) {
+	jsigner, err := jose.NewSigner(
+		jose.SigningKey{
+			Algorithm: jose.EdDSA,
+			Key:       c.clientSecret,
+		},
+		&jose.SignerOptions{
+			NonceSource: &xjwt.RandomNonce{Size: 16},
+		})
+	if err != nil {
+		return nil, err
+	}
+
+	// Our token host may include a port, but the audience never expects a port
+	aud := c.tokenHost
+	host, _, err := net.SplitHostPort(aud)
+	if err == nil {
+		aud = host
+	}
+	now := time.Now()
+	claims := &jwt.Claims{
+		Issuer:    c.clientID,
+		Subject:   c.clientID,
+		Audience:  jwt.Audience{aud},
+		Expiry:    jwt.NewNumericDate(now.Add(time.Minute * 2)),
+		IssuedAt:  jwt.NewNumericDate(now),
+		NotBefore: jwt.NewNumericDate(now.Add(-time.Minute * 2)),
+	}
+	b, err := json.Marshal(claims)
+	if err != nil {
+		return nil, err
+	}
+
+	rv, err := jsigner.Sign(b)
+	if err != nil {
+		return nil, err
+	}
+
+	s, err := rv.CompactSerialize()
+	if err != nil {
+		return nil, err
+	}
+
+	body := url.Values{
+		"client_id":             []string{c.clientID},
+		"grant_type":            []string{"client_credentials"},
+		"client_assertion_type": []string{assertionType},
+		"client_assertion":      []string{s},
+	}
+
+	tokenHost := c.tokenHost
+	if envHost, ok := os.LookupEnv("BATON_C1_API_HOST"); ok {
+		tokenHost = envHost
+	}
+	tokenUrl := url.URL{
+		Scheme: "https",
+		Host:   tokenHost,
+		Path:   "auth/v1/token",
+	}
+
+	req, err := http.NewRequestWithContext(context.Background(), http.MethodPost, tokenUrl.String(), strings.NewReader(body.Encode()))
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	resp, err := http.DefaultClient.Do(req)
```

Comment:
could this use our uhttp client?  defaultclient has... tradeoffs? no user-agent?

Reply:
yeah actually the struct has an http.Client, should use that

Reply:
Good call. We were already creating a client to use for other calls, just forgot to update this one. Fixed in cfad197.